### PR TITLE
Removing configurability of `Jenkins.agentProtocols`

### PR DESCRIPTION
**Pull request by jglick**
I ran into a case where a CloudBees CI customer was missing a proprietary agent protocol (see #8621). It seems that this was quite inadvertent: they had in the past exported a JCasC bundle that included something like

```yaml
jenkins:
  agentProtocols:
  - Ping
  - JNLP4-connect
```

despite never having touched this configuration section, and without paying attention to this line had committed it to SCM, where it prevented new protocols from appearing! (You can find lots of similar examples, some outdated, [by searching](https://github.com/search?type=code&q=org%3Ajenkinsci+agentprotocols+language%3AYAML).)

The history here is long and tortured (#2010, #2950, #3188, #4387, #7451, etc.) but essentially there _was_ a period of Jenkins development in which new TCP (agent) protocols were being actively developed, and a rather complex opt-in-and-out system was devised to allow administrators to pick which protocol(s) to allow and have that choice be partially preserved across upgrades. But for years now, there has only been `JNLP4-connect`, and there is no meaningful choice: if you do not wish to allow inbound TCP agents, just disable the port.

The entire protocol selection system seems like technical debt now. This PR just deletes it all and treats `AgentProtocol` like any other headless extension: whichever protocols are registered by core or plugins are allowed. If a plugin developer wishes to conditionally roll out a protocol behind a feature flag, they can conditionally return `null` from `getName`. If an administrator has some weird reason to disable one protocol (but allow others) they could [filter it out](https://plugins.jenkins.io/extension-filter/), but this is not something we would advertise as a supported controller customization.

### Testing done

The protocols section in **Configure Security** is gone.

### Proposed changelog entries

- Stop allowing configuration of the list of agent protocols.

### Proposed upgrade guidelines

The list of agent protocols is no longer configurable, either through the GUI or JCasC. (Normally there will be two protocols, one for inbound TCP agents, and a ping protocol used to test connectivity.) The `agentProtocols` section of a JCasC bundle will now be ignored and should be deleted. If you do not wish to allow inbound TCP agents, disable the port rather than the protocol.

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/resources/jenkins/slaves/Messages.properties
Comment (jglick) [inline core/src/main/resources/jenkins/slaves/Messages.properties:25]: Apparently forgotten in some earlier cleanup.

File: core/src/main/resources/jenkins/slaves/Messages.properties
Comment (jglick) [inline core/src/main/resources/jenkins/slaves/Messages.properties:27]: Ditto.

**Review by timja — APPROVED** [review]
Makes sense, thanks!

**Review by Vlatombe — APPROVED** [review]
LGTM

**Comment by Vlatombe** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by MarkEWaite** [comment]
This causes test failures in the plugin BOM.  The configuration as code plugin shows several test failures with this change.  Other plugins may also show failures.  I've started a plugin BOM run to check other plugins:

* https://github.com/jenkinsci/bom/pull/3847

@jglick would you be willing to submit the pull request to the configuration as code plugin to adapt its tests to this change?

**Comment by jglick** [comment]
Sorry, I forgot to mark those two ready for review after this merged. Did that now. CC @timja & @Dohbedoh as likely maintainers.

**Comment by MarkEWaite** [comment]
Thanks @jglick !  The incremental builds of those plugins passed the plugin BOM tests in

* https://github.com/jenkinsci/bom/pull/3847

I would love to have a release of those two plugins at about the same time as 2.483 releases tomorrow so that the plugin BOM can include those new releases.

**Comment by jglick** [comment]
Well https://github.com/jenkinsci/configuration-as-code-plugin/pull/2580 is outside my control. https://github.com/jenkinsci/support-core-plugin/pull/592 I could merge myself but would prefer to wait for a review from _somebody_.

**Comment by amuniz** [comment]
Could this be breaking additional protocols defined by proprietary software (I know at least of one custom protocol defined by CloudBees for network diagnosis).

**Comment by jglick** [comment]
No, I do not think so.

**Comment by Zhenye-Na** [comment]
Hello there,

I'd like to bring attention to a potential backward compatibility issue observed with the Jenkins 2.492.1 release (https://www.jenkins.io/changelog/2.492.1/). 

Upon upgrading, we encountered the following error:

```
SEVERE	jenkins.InitReactorRunner$1#onTaskFailed: Failed ConfigurationAsCode.init

io.jenkins.plugins.casc.ConfiguratorException: 'agentProtocols' is deprecated
```

The server logs indicate an upgrade attempt:


```
INFO	hudson.PluginManager#loadDetachedPlugins: Upgrading Jenkins. The last running version was 2.479.3. This Jenkins is version 2.492.1.
```

After reviewing the release changelog, it appears that the `agentProtocols` field should no longer be used. This change seems to be causing issues with existing configurations.

Given this situation, I'd like to inquire about the current process for ensuring backward compatibility in releases. Are there plans to enhance compatibility testing procedures for future releases?

Thank you for your attention to this matter.

**Comment by MarkEWaite** [comment]
> Given this situation, I'd like to inquire about the current process for ensuring backward compatibility in releases. Are there plans to enhance compatibility testing procedures for future releases?

There are no plans to change that behavior of configuration as code as far as I know.  It is expected that administrators will read the changelog, read the upgrade guide, and heed the instructions in those documents.  When configuration as code is using an invalid configuration (with a key that is no longer available), Jenkins stops rather than making a potentially flawed assumption about what the administrator intends based on the incorrect definition in the configuration as code definition.

Stopping immediately seems to me like the best way to handle the issue.  If Jenkins continues running with the incorrect configuration, then administrators that ignore startup warnings will accumulate more and more cases where their incorrect configuration as code definition is still being used.

**Comment by jglick** [comment]
https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/PLUGINS.md#backward-compatibility-test claims you can request that deprecated attributes not break startup. I have not tried it. In general the expectation is that a given version of a configuration bundle needs to be tested against a specific core/plugin version combination and may occasionally need adjustment. This is in contrast to GUI-configured XML settings, for which we strive to maintain automated backward compatibility across upgrades.

**Comment by cpotteck** [comment]
@MarkEWaite 

> It is expected that administrators will read the changelog, read the upgrade guide, and heed the instructions in those documents. When configuration as code is using an invalid configuration (with a key that is no longer available), Jenkins stops rather than making a potentially flawed assumption about what the administrator intends based on the incorrect definition in the configuration as code definition.

This could be clarified in the ["Upgrading to Jenkins LTS 2.492.x" guide](https://www.jenkins.io/doc/upgrade-guide/2.492/#upgrading-to-jenkins-lts-2-492-1), as it currently reads as if it will simply be ignored, rather than failing at startup:

> The agentProtocols section of a JCasC bundle should be deleted because it will now be ignored.

**Comment by MarkEWaite** [comment]
I've submitted a pull request to provide more details in the upgrade guide:

* https://github.com/jenkins-infra/jenkins.io/pull/7853

**Pull request merged**
